lib/thor/parser/arguments.rb

```diff
@@ -122,7 +122,15 @@ def parse_hash(name)
     def parse_array(name)
       return shift if peek.is_a?(Array)
       array = []
-      array << shift while current_is_value?
+      while current_is_value?
+        value = shift
+        if !value.empty? && @switches.is_a?(Hash) && switch = @switches[name]
```

Comment:
Can we restructure this conditional a bit? The `=` towards the end is kind of like a shark in calm waters, it's really unexpected.